tests/api/test_remote_engine.py

```diff
@@ -207,3 +208,164 @@ def test_default_compiler(self, prog, monkeypatch, caplog):
 
         assert engine.device_spec.default_compiler == "Xunitary"
         assert caplog.records[-1].message == "Compiling program for device X8_01 using compiler Xunitary."
+
+    class MockProgram:
+        """A mock program for testing"""
+        def __init__(self):
+            self.run_options = {}
+
+    def test_no_compile_info(self, prog, monkeypatch, caplog):
+        """Tests that compilation happens by default if no compile_info was
+        specified when call run_async."""
+        caplog.set_level(logging.INFO)
+        test_device_dict = mock_device_dict.copy()
+        test_device_dict["compiler"] = []
+
+        monkeypatch.setattr(Connection, "create_job", lambda self, target, program, run_options: program)
+        monkeypatch.setattr(Connection, "_get_device_dict", lambda *args: test_device_dict)
+        monkeypatch.setattr(Program, "compile", lambda *args, **kwargs: self.MockProgram())
+
+        # Leaving compile_info as None
+        assert prog.compile_info == None
+
+        engine = RemoteEngine("X8")
+        program = engine.run_async(prog, shots=10)
+
+        assert isinstance(program, self.MockProgram)
+        assert caplog.records[-1].message == "Compiling program for device X8_01 using compiler Xunitary."
+
+    def test_compile_device_invalid_device_error(self, prog, monkeypatch, caplog):
+        """Tests that an error is raised if the program was compiled for
+        another device and recompilation was not requested."""
+        caplog.set_level(logging.INFO)
+        test_device_dict = mock_device_dict.copy()
+        test_device_dict["compiler"] = []
+
+        monkeypatch.setattr(Connection, "create_job", lambda self, target, program, run_options: program)
+        monkeypatch.setattr(Connection, "_get_device_dict", lambda *args: test_device_dict)
+        monkeypatch.setattr(Program, "compile", lambda *args, **kwargs: self.MockProgram())
+
+        # Setting compile_info with a dummy devicespec and compiler name
+        X8_spec = DeviceSpec(target="DummyDevice", connection=None, spec=None)
+        prog._compile_info = (X8_spec, "dummy_compiler")
+
+        engine = sf.RemoteEngine("X8")
+        with pytest.raises(
+            ValueError, match="Cannot use program compiled for"
+        ):
+            program = engine.run_async(prog, shots=10)
+
+    def test_recompile_different_device(self, prog, monkeypatch, caplog):
+        """Tests that no error is raised if the program was compiled for
```

Comment:
What exactly does this mean? It's not immediately obvious to me how a program compiled for one device could be recompiled for another.

Reply:
What is meant to be tested here, is that even if the program was compiled for a device that is not compatible with the default device specification, since `recompile=True`, recompilation would take place anyways.

Reworded the docstring, let me know if it would be better to further add details here.

Reply:
Got it, thanks!